contributors/design-proposals/sig-cli/dynamic_resource_shortcut_fecth.md

```diff
@@ -0,0 +1,107 @@
+
+For how to prompt those information properly, there're two solutions here I've come up with:
+
+### solution 1
+
+Keep current help messages, and add a ring1Factory function use clientDiscovery fetch resources/shortcuts from server then concatenate those valid resources string and print every time when user execute command,also if cluster is not aviliable for some reason, kubectl will fall back to use hardcode resource strings.
+
+```
+➜  ~ kubectl get -h
+Display one or many resources.
+
+Valid resource types include: 
+
+  * all  
+  * certificatesigningrequests (aka 'csr')  
+  * clusterrolebindings  
+  * clusterroles  
+  * clusters (valid only for federation apiservers)  
+  * componentstatuses (aka 'cs')  
+  * configmaps (aka 'cm')  
+  ...
+```
+
+### solution 2
+
+Remove those valid resources list in help messages, and add a dedicated command to fetch resources/shortcuts, like `kubectl api-resources`,implement pr [here](https://github.com/kubernetes/kubernetes/pull/42873) then tell user use `kubectl api-resources` to check what kind of valid resouces they can use.
+
+```
+➜  ~ kubectl get -h
+Display one or many resources. 
+
+You can use `kubectl api-resources` to check valid resource types server provided. 
+
+...
+```
+
+
+## User Experience
+
+### Use Cases
+
+`kubectl get/describe/explain..   -h`
+
+
+### Shortage
+
+solution 1 may have some shortage in first time prompt help messages.
+
+    ResourceAliasesResourceShortFormForkubectl get -h  0.13s user 0.02s system 128% cpu 0.117 total
+    ./kubectl get -h  0.13s user 0.02s system 5% cpu 2.600 total
+
+
+
+## Implementation
+
+### solution 1
+
+Add function/interface in ring1Factory  `func ValidResourcesFromDiscoveryClient []ResourceShortcut`
+
+change for resource in kubectl get/describe/explain... -h
+
+    pkg/kubectl/cmd/get.go
+    cmd := &cobra.Command{
+        ...
+        Short:   i18n.T("Display one or many resources"),
+        Long:    templates.LongDynamicDesc(getLong, f.ValidResourcesFromDiscoveryClient())  
+        Example: getExample,
+
+    pkg/kubectl/cmd/templates/normalizers.go    
+    func LongDynamicDesc(s string, r []kubectl.ResourceShortcuts) {
+    	if len(r) != 0 {
+    		//foreach ResourcesShortcuts sorted, and concatenate to string  
+    		return LongDesc(fmt.Sprintf(getLong, concatenateString))
+    	}
+    	return LongDesc(fmt.Sprintf(getLong, validResours))
+    }
+
+
+### solution 2 
+
+This is a simple one, only change is remove hardcoded list in help messages of `kubectl get/explain/describe ...` and implement a new command `kubectl api-resources`.
+
+
+## Client/Server Backwards/Forwards compatibility
+
+Depend on which solution we choose.
+
+## Other things need to discuss
+
+1. Use binary cache from OpenAPI instead of cache from `.kube/cache/discovery` ? I think this is only happend in we chose solution one, and I think openapi is more suit for validation, and the information we need here is only resources name and shortcut. 
```

Comment:
@liggitt Can we actually get that data from OpenAPI? I don't think there is any "shortcut" in OpenAPI.